scala-xml/src/main/scala/org/http4s/scalaxml/ElemInstances.scala

```diff
@@ -78,17 +79,21 @@ trait ElemInstances {
       val source = new InputSource()
       msg.charset.foreach(cs => source.setEncoding(cs.nioCharset.name))
 
-      collectBinary(msg).flatMap[DecodeFailure, Elem] { chunk =>
-        source.setByteStream(new ByteArrayInputStream(chunk.toArray))
-        val saxParser = saxFactory.newSAXParser()
-        EitherT(
-          F.blocking(XML.loadXML(source, saxParser))
-            .map(Either.right[DecodeFailure, Elem](_))
-            .recover { case e: SAXParseException =>
-              Left(MalformedMessageBodyFailure("Invalid XML", Some(e)))
-            }
-        )
-      }
+      EitherT(
+        msg.body
+          .through(toInputStream)
+          .evalMap { in =>
+            source.setByteStream(in)
+            val saxParser = saxFactory.newSAXParser()
+            F.blocking(XML.loadXML(source, saxParser))
+          }
+          .compile
+          .lastOrError
```

Comment:
I think you could also use `toInputStreamResource` here instead and save the awkward `lastOrError`.

Reply:
Oh, smart.

Reply:
Nice, cleaned up nicely!

Reply:
The property test (disclaimer: not a benchmark) got about 30% faster, too.